File: src/window-state.js
// Место окна между запусками (1.1.0). Тот же модуль, что в калькуляторах
// (Calculators: calcpro-glass/window-state.js) и в Paint Pro; у SyncGlass окно
// постоянного размера, так что запоминается только место.
//
// Правило выбора места - чистые функции без Electron, их проверяют обычные тесты;
// main.js только читает и пишет файл и берёт у Electron рабочие области экранов.
// Что может прийти из файла: окно на отключённом мониторе, размер больше экрана или
// меньше минимума, мусор вместо чисел, пустой или обрезанный файл. Всё это должно
// давать окно, которое видно и за которое можно взяться, а не падение.
'use strict';

const fs = require('fs');
const path = require('path');

// Сколько окна должно оставаться на экране, чтобы за него можно было взяться: полоса
// заголовка высотой 38 (системный заголовок Windows с запасом) и хотя бы 80 по ширине.
const GRIP_HEIGHT = 38;
const GRIP_WIDTH = 80;

const finite = (v) => typeof v === 'number' && Number.isFinite(v);

function overlap(a, b) {
  const w = Math.min(a.x + a.width, b.x + b.width) - Math.max(a.x, b.x);
  const h = Math.min(a.y + a.height, b.y + b.height) - Math.max(a.y, b.y);
  return w > 0 && h > 0 ? { width: w, height: h } : null;
}

/**
 * Где и какого размера открыть окно.
 * @param saved   что лежало в файле (что угодно, в том числе null)
 * @param areas   рабочие области экранов {x, y, width, height}, основной первым
 * @param opts    {width, height, minWidth, minHeight} - размер по умолчанию и минимум
 * @returns {{width, height, x?, y?, maximized}} без x и y - Electron поставит окно по центру
 */
function restore(saved, areas, opts) {
  const byDefault = { width: opts.width, height: opts.height, maximized: false };
  const screens = (Array.isArray(areas) ? areas : []).filter((a) =>
    a && finite(a.x) && finite(a.y) && finite(a.width) && finite(a.height) && a.width > 0 && a.height > 0);
  if (!saved || typeof saved !== 'object' || !finite(saved.width) || !finite(saved.height)) return byDefault;

  const maximized = saved.maximized === true;
  // Самый большой экран ограничивает размер сверху: окно, сохранённое на большом
  // мониторе, не должно открыться на маленьком шире экрана.
  const biggest = screens.reduce((m, a) => ({ width: Math.max(m.width, a.width), height: Math.max(m.height, a.height) }),
    { width: opts.width, height: opts.height });
  let width = Math.round(Math.min(Math.max(saved.width, opts.minWidth), Math.max(biggest.width, opts.minWidth)));
  let height = Math.round(Math.min(Math.max(saved.height, opts.minHeight), Math.max(biggest.height, opts.minHeight)));

  if (!finite(saved.x) || !finite(saved.y)) return { width, height, maximized };
  const x = Math.round(saved.x), y = Math.round(saved.y);

  // Экран, на котором больше всего полосы заголовка. Если заголовок не виден ни на одном
  // (монитор отключили, разрешение уменьшили) - окно по центру, размер сохраняется.
  const grip = { x, y, width, height: GRIP_HEIGHT };
  let best = null, bestArea = 0;
  for (const a of screens) {
    const o = overlap(grip, a);
    if (o && o.width * o.height > bestArea) { best = a; bestArea = o.width * o.height; }
  }
  if (!best || bestArea < GRIP_WIDTH * GRIP_HEIGHT / 2) return { width, height, maximized };

  // Окно, чуть съехавшее за край, придвигается к краю своего экрана целиком. Если оно
  // выше экрана (экран ниже минимума окна), прижимается верхом: заголовок важнее низа.
  // Прижимать низом было ошибкой - заголовок уезжал выше экрана (нашло свойство
  // «восстановленное восстанавливается без изменений» на экране 640x481).
  width = Math.min(width, Math.max(best.width, opts.minWidth));
  height = Math.min(height, Math.max(best.height, opts.minHeight));
  const nx = Math.max(best.x, Math.min(x, best.x + best.width - width));
  const ny = Math.max(best.y, Math.min(y, best.y + best.height - height));
  return { x: nx, y: ny, width, height, maximized };
}

/** Что запомнить: обычные границы окна (и у развёрнутого - те, к которым оно вернётся). */
function capture(win) {
  const b = typeof win.getNormalBounds === 'function' ? win.getNormalBounds() : win.getBounds();
  return { x: b.x, y: b.y, width: b.width, height: b.height, maximized: win.isMaximized() };
}

/** Прочитать файл; нет файла или в нём мусор - null. */
function load(file) {
  try {
    return JSON.parse(fs.readFileSync(file, 'utf8'));
  } catch {
    return null;
  }
}

/**
 * Записать файл: сначала во временный, потом переименовать. Если процесс убьют на
 * середине записи, останется старый файл, а не обрезанный.
 */
function save(file, state) {
  try {
    fs.mkdirSync(path.dirname(file), { recursive: true });
    const tmp = file + '.tmp';
    fs.writeFileSync(tmp, JSON.stringify(state));
    fs.renameSync(tmp, file);
    return true;
  } catch {
    return false; // не записалось - в следующий раз окно просто откроется по умолчанию
  }
}

module.exports = { restore, capture, load, save, GRIP_HEIGHT, GRIP_WIDTH };

File: csharp/src/SyncGlass.Core/WindowState.cs
using System.Text.Json.Nodes;

namespace SyncGlass.Core;

/// <summary>
/// Перенос src/window-state.js: место окна между запусками. Тот же модуль, что в
/// калькуляторах и Paint Pro; у SyncGlass окно постоянного размера, так что по сути
/// запоминается место. Файл window-state.json общий с Electron-версией:
/// {x, y, width, height, maximized}.
///
/// Что может прийти из файла: окно на отключённом мониторе, размер больше экрана или
/// меньше минимума, мусор вместо чисел, пустой или обрезанный файл. Всё это должно
/// давать окно, которое видно и за которое можно взяться, а не падение.
/// </summary>
public static class WindowState
{
    // Сколько окна должно оставаться на экране, чтобы за него можно было взяться: полоса
    // заголовка высотой 38 (системный заголовок Windows с запасом) и хотя бы 80 по ширине.
    public const double GripHeight = 38;
    public const double GripWidth = 80;

    public readonly record struct Area(double X, double Y, double Width, double Height);

    // Что лежало в файле: любое поле может оказаться не числом - тогда null.
    public sealed record Saved(double? X, double? Y, double? Width, double? Height, bool Maximized);

    // Где и какого размера открыть окно; X и Y = null - поставить по центру.
    public sealed record Placement(double? X, double? Y, double Width, double Height, bool Maximized);

    public sealed record Options(double Width, double Height, double MinWidth, double MinHeight);

    private static bool Finite(double? v) => v is { } d && double.IsFinite(d);

    public static Placement Restore(Saved? saved, IEnumerable<Area> areas, Options opts)
    {
        var byDefault = new Placement(null, null, opts.Width, opts.Height, false);
        var screens = areas.Where(a => double.IsFinite(a.X) && double.IsFinite(a.Y) && double.IsFinite(a.Width)
                                       && double.IsFinite(a.Height) && a.Width > 0 && a.Height > 0).ToList();
        if (saved == null || !Finite(saved.Width) || !Finite(saved.Height)) return byDefault;

        var maximized = saved.Maximized;
        // Самый большой экран ограничивает размер сверху.
        var bw = screens.Aggregate(opts.Width, (m, a) => Math.Max(m, a.Width));
        var bh = screens.Aggregate(opts.Height, (m, a) => Math.Max(m, a.Height));
        var width = JsRound(Math.Min(Math.Max(saved.Width!.Value, opts.MinWidth), Math.Max(bw, opts.MinWidth)));
        var height = JsRound(Math.Min(Math.Max(saved.Height!.Value, opts.MinHeight), Math.Max(bh, opts.MinHeight)));

        if (!Finite(saved.X) || !Finite(saved.Y)) return new Placement(null, null, width, height, maximized);
        var x = JsRound(saved.X!.Value);
        var y = JsRound(saved.Y!.Value);

        // Экран, на котором больше всего полосы заголовка. Не видна ни на одном - по центру.
        Area? best = null;
        double bestArea = 0;
        foreach (var a in screens)
        {
            var ow = Math.Min(x + width, a.X + a.Width) - Math.Max(x, a.X);
            var oh = Math.Min(y + GripHeight, a.Y + a.Height) - Math.Max(y, a.Y);
            if (ow > 0 && oh > 0 && ow * oh > bestArea)
            {
                best = a;
                bestArea = ow * oh;
            }
        }
        if (best is not { } b || bestArea < GripWidth * GripHeight / 2) return new Placement(null, null, width, height, maximized);

        // Съехавшее окно придвигается к краю своего экрана целиком; выше экрана - верхом:
        // заголовок важнее низа.
        width = Math.Min(width, Math.Max(b.Width, opts.MinWidth));
        height = Math.Min(height, Math.Max(b.Height, opts.MinHeight));
        var nx = Math.Max(b.X, Math.Min(x, b.X + b.Width - width));
        var ny = Math.Max(b.Y, Math.Min(y, b.Y + b.Height - height));
        return new Placement(nx, ny, width, height, maximized);
    }

    // Math.round из JS: половина - вверх (к +∞), а не к чётному, как Math.Round.
    private static double JsRound(double v) => Math.Floor(v + 0.5);

    // Прочитать файл; нет файла или в нём не объект - null. Поле не числом - null.
    public static Saved? Load(string file)
    {
        try
        {
            if (JsonNode.Parse(File.ReadAllText(file)) is not JsonObject o) return null;
            static double? Num(JsonNode? n) => n is JsonValue v && v.TryGetValue<double>(out var d) ? d : null;
            var max = o["maximized"] is JsonValue m && m.TryGetValue<bool>(out var bm) && bm;
            return new Saved(Num(o["x"]), Num(o["y"]), Num(o["width"]), Num(o["height"]), max);
        }
        catch
        {
            return null;
        }
    }

    // Записать через временный файл: убьют на середине - останется старый, а не обрезанный.
    public static bool Save(string file, Placement p)
    {
        try
        {
            Directory.CreateDirectory(Path.GetDirectoryName(file)!);
            var tmp = file + ".tmp";
            var o = new JsonObject();
            if (p.X is { } x) o["x"] = x;
            if (p.Y is { } y) o["y"] = y;
            o["width"] = p.Width;
            o["height"] = p.Height;
            o["maximized"] = p.Maximized;
            File.WriteAllText(tmp, o.ToJsonString());
            FsOps.Rename(tmp, file);
            return true;
        }
        catch
        {
            return false; // не записалось - в следующий раз окно откроется по умолчанию
        }
    }
}
